src/coreclr/nativeaot/System.Private.StackTraceMetadata/src/Internal/StackTraceMetadata/StackTraceMetadata.cs

```diff
@@ -101,12 +107,91 @@ public static unsafe string GetMethodNameFromStartAddressIfAvailable(IntPtr meth
+                return methodName;
             }
 
+            owningTypeName = null;
+            genericArgs = null;
+            methodSignature = null;
+            hashCodeForLineInfo = 0;
             return null;
         }
 
+        private static unsafe (string, int) GetLineNumberInfo(IntPtr methodStartAddress, int offset, int hashCode)
+        {
+            foreach (NativeFormatModuleInfo module in ModuleList.EnumerateModules())
+            {
+                if (!module.TryFindBlob((int)ReflectionMapBlob.BlobIdStackTraceLineNumbers, out byte* lineNumbersBlob, out uint cbLineNumbersBlob)
+                    || !module.TryFindBlob((int)ReflectionMapBlob.BlobIdStackTraceDocuments, out byte* documentsBlob, out uint cbDocumentsBlob))
+                {
+                    continue;
+                }
+
+                ExternalReferencesTable externalReferences = default(ExternalReferencesTable);
+                externalReferences.InitializeCommonFixupsTable(module);
+
+                var reader = new NativeReader(lineNumbersBlob, cbLineNumbersBlob);
+                var parser = new NativeParser(reader, 0);
+                var hashtable = new NativeHashtable(parser);
+                var lookup = hashtable.Lookup(hashCode);
+                NativeParser entryParser;
+
+                while (!(entryParser = lookup.GetNext()).IsNull)
+                {
+                    IntPtr foundMethod = externalReferences.GetFunctionPointerFromIndex(entryParser.GetUnsigned());
+                    if (foundMethod != methodStartAddress)
+                        continue;
+
+                    uint numEntries = entryParser.GetUnsigned();
+
+                    int currentLineNumber = 0;
+                    int currentNativeOffset = 0;
+                    int documentIndex = entryParser.GetSigned();
+
+                    for (uint i = 0; i < numEntries; i++)
+                    {
+                        int nativeOffsetDelta = entryParser.GetSigned();
+                        int oldDocumentIndex = documentIndex;
+                        if (i > 0 && nativeOffsetDelta == 0)
+                        {
+                            documentIndex = entryParser.GetSigned();
+                            nativeOffsetDelta = entryParser.GetSigned();
+                        }
+
+                        if (currentNativeOffset + nativeOffsetDelta > offset)
+                        {
+                            documentIndex = oldDocumentIndex;
+                            break;
+                        }
+
+                        int lineNumberDelta = entryParser.GetSigned();
+
+                        currentNativeOffset += nativeOffsetDelta;
+                        currentLineNumber += lineNumberDelta;
+                    }
+
+                    int documentOffset = ((int*)documentsBlob)[documentIndex];
```

Comment:
```suggestion
                    int documentOffset = BinaryPrimitives.ReadInt32LittleEndian(new ReadOnlySpan(documentsBlob, cbDocumentsBlob).Slice(documentIndex * sizeof(int)));
```
Writer is little endian, reader should match

Reply:
> Writer is little endian, reader should match

Wouldn't it be suboptimal? If we ever start caring about big endian, we should emit this as big endian. These data structures are native AOT - and big endian is the native thing for those arches. We don't currently have a "is the target big endian" as a concept in the compiler.

Reply:
> The endianness swaps would be problematic for pointers (e.g. vtable slots - we would have some work to do for these if we started caring about big endian).

Yep, that's my thinking. E.g. we wouldn't generate the numVTableSlots field on MethodTable as little endian. We could probably make it work with little endian. It would not be nice.

Since we haven't really started caring about big endian in any of our data structures, I'll leave this to someone in the future who does (likely not on Microsoft payroll).